cordapp-example/contracts-kotlin/src/main/kotlin/com/example/contract/IOUContract.kt

```diff
@@ -41,23 +40,19 @@ class IOUContract : Contract {
                 val out = tx.outputsOfType<IOUState>().single()
                 "The lender and the borrower cannot be the same entity." using (out.lender != out.borrower)
                 "Command require two signers." using (command.signers.size == 2)
-                "All of the participants must be signers." using (command.signers.containsAll(out.participants.map { it.owningKey }))
+                "All of the participants must be signers." using (command.signers.containsAll(listOf(out.lender.owningKey, out.borrower.owningKey)))
 
                 // IOU-specific constraints.
                 "The IOU's value must be non-negative." using (out.value > 0)
             }
             is Commands.Destroy -> requireThat {
-                "There should be two inputs in general." using (tx.inputs.size == 2)
-                "There should be only one IOUState input." using (tx.inputsOfType<IOUState>().size == 1)
-                "There should be only one CashState input." using (tx.inputsOfType<CashState>().size == 1)
+                "There should be one input." using (tx.inputsOfType<IOUState>().size == 1)
+                "There should be no output." using tx.outputsOfType<IOUState>().isEmpty()
 
-                "There should be one output in general." using (tx.outputs.size == 1)
-                "There should be one CashState output." using (tx.outputsOfType<CashState>().size == 1)
-
-                val iouStateSigner = tx.inputsOfType<IOUState>().single().lender.owningKey
+                val lenderPK = tx.inputsOfType<IOUState>().single().lender.owningKey
 
                 "Expect one signer." using (command.signers.size == 1)
-                "Lender must be a signer." using (command.signers.single() == iouStateSigner)
+                "Lender must be a signer." using (command.signers.single() == lenderPK)
```

Comment:
Thought experiment here.

1. Do you absolutely want the lender to sign?
2. Or do you absolutely want nobody to sign other than the signer?

What I am driving at is, I would go with 1: `using (command.signers.contains(lenderPK))`. Because who knows, a flow may mistakenly add the borrower in the signers, but the contract says "does not matter as long as I see the lender".

Reply:
ok, good point, but @adelrustum bring other side too my attention also: if I expect one signature and there will be more... isn't that suspicious? if the flow add something by mistake, isn't that a bug that should be fixed?

taking both sides into consideration I have one question: is there any danger that can be done by additional signature that is not expected? I can't thing about anything other that: some other party has access to transaction data.

Reply:
I oscillate between the 2 :).

The risk is when a flow adds an extra signer, this extra signer will be asked to sign. Normally, it may not need to sign, so this should be flagged, so point 1 is safe-ish. But are you sure you flag properly things you are asked to sign but should not?

In practice, when asked to sig, it is for the whole transaction so if you are asked to sign "only for this command", in fact it is for the whole tx, which may contain nefarious things. We come back to "check thoroughly before you sign".

Reply:
There's no harm of someone `irrelevant` signing your transaction; the problem is when a `required` signature is missing, it means someone who's supposed to approve this transaction didn't give their approval.

Reply:
Ah yes, you are introducing a party that could hold up the flow "no I am not signing this tx that has nothing to do with me", all because you mistakenly add them as a required signer.